fn/context_guard_test.go

```diff
@@ -0,0 +1,434 @@
+			g           = NewContextGuard()
+		)
+
+		ctxc, _ := g.WithCtx(ctx)
+
+		// Cancel the parent context.
+		cancel()
+		// Assert that the derived context is cancelled.
+		select {
+		case <-ctxc.Done():
+		default:
+			t.Errorf("The derived context should be cancelled at " +
+				"this point")
+		}
+	})
+
+	// Test that the derived context is cancelled when the returned cancel
+	// function is called.
+	t.Run("Derived context is cancelled", func(t *testing.T) {
+		t.Parallel()
+		var (
+			ctx = context.Background()
+			g   = NewContextGuard()
+		)
+
+		ctxc, cancel := g.WithCtx(ctx)
+
+		// Cancel the context.
+		cancel()
+
+		// Assert that the derived context is cancelled.
+		select {
+		case <-ctxc.Done():
+		default:
+			t.Errorf("The derived context should be cancelled at " +
+				"this point")
+		}
+	})
+
+	// Test that the derived context is cancelled when the quit channel is
+	// closed.
+	t.Run("Quit channel is closed", func(t *testing.T) {
+		t.Parallel()
+
+		var (
+			ctx = context.Background()
+			g   = NewContextGuard()
+		)
+		ctxc, _ := g.WithCtx(ctx)
+
+		// Close the quit channel.
+		g.Quit()
+
+		// Assert that the derived context is cancelled.
+		select {
+		case <-ctxc.Done():
+		default:
+			t.Errorf("The derived context should be cancelled at " +
+				"this point")
+		}
+	})
+
+	t.Run("Parent context is already closed", func(t *testing.T) {
+		t.Parallel()
+
+		var (
+			ctx, cancel = context.WithCancel(context.Background())
+			g           = NewContextGuard()
+		)
+		cancel()
+
+		ctxc, _ := g.WithCtx(ctx)
+		// Assert that the derived context is cancelled already
+		// cancelled.
+		select {
+		case <-ctxc.Done():
+		default:
+			t.Errorf("The derived context should be cancelled at " +
+				"this point")
+		}
+	})
+
+	t.Run("Quit channel is already closed", func(t *testing.T) {
+		t.Parallel()
+
+		var (
+			ctx = context.Background()
+			g   = NewContextGuard()
+		)
+
+		g.Quit()
+
+		ctxc, _ := g.WithCtx(ctx)
+
+		// Assert that the derived context is cancelled already
+		// cancelled.
+		select {
+		case <-ctxc.Done():
+		default:
+			t.Errorf("The derived context should be cancelled at " +
+				"this point")
+		}
+	})
+
+	t.Run("Child context should be cancelled synchronously with "+
+		"parent", func(t *testing.T) {
+
+		t.Parallel()
+
+		var (
+			ctx, cancel = context.WithCancel(context.Background())
+			g           = NewContextGuard()
+			task        = make(chan struct{})
+			done        = make(chan struct{})
+		)
+		// Derive a child context.
+		ctxc, _ := g.WithCtx(ctx)
+
+		// Spin off a routine that exists cleaning if the child context
+		// is cancelled but fails if the task is performed.
+		go func() {
+			defer close(done)
+			select {
+			case <-ctxc.Done():
+			case <-task:
+				t.Fatalf("should not get here")
+			}
+		}()
+
+		// Give the goroutine above a chance to spin up so that it's
+		// waiting on the select.
+		time.Sleep(time.Millisecond * 200)
```

Comment:
nit: alternatively you could read from a temp channel that is written to in the goroutine's body.

Reply:
(and below for similar sleeps).

Reply:
if we do that then yes we know that the goroutine has at least started, but we dont actually know that it is hanging on the `select` so it will still be flaky in terms of timing there